text/0000-impl-enum.md

```diff
@@ -0,0 +1,341 @@
+enum Enum {
+    Variant1,
+    Variant2(...),
+    Variant3{...},
+    ...
+}
+```
+
+Now enums implicitly generate structs for each variant like this:
+
+```rust
+mod Enum {
+    pub struct Variant1;
+    pub struct Variant2(...);
+    pub struct Variant3{...};
+    ...
+}
+```
+
+That's not the main feature of this proposal, but necessary to get it work.
+
+Additinally, enums generate a type similar to traits for the enum, which are implemented by the structs, and also work in a similar way.
+The current enums are the counterpart to trait objects then.
+
+# Motivation
+[motivation]: #motivation
+
+Let's begin with an example:
+
+```rust
+use recieve_exit_event; // fn() -> bool
+
+trait Updateable<E> {
+    fn update(&self, event: E) -> bool;
+}
+
+enum Event {
+    Idle,
+    Exit,
+}
+
+struct Object;
+impl Updateable<Event> for Object {
+    fn update(&self, event: Event) -> bool {
+        match event {
+            Event::Idle => true,
+            Event::Exit => false,
+        }
+    }
+}
+
+fn run() {
+    let current_updatable = (&Object as &dyn Updateable<Event>);
+    loop {
+        if recieve_exit_event() {
+            if !current_updatable.update(Event::Exit) {
+                break;
+            }
+        }
+        if !current_updatable.update(Event::Idle) {
+            break;
+        }
+    }
+}
+```
+
+This is a small implementation of an event loop. Inside the loop, new events are generated under certain conditions. Then they are passed to a method, which handles these events.
+
+The event types will probably contain some information important for the events.
+
+The event types are known at compile time, so it would be more efficient to write it like this:
+
+```rust
+use recieve_exit_event; // fn() -> bool
+
+trait Updateable<IdleEvent, ExitEvent> {
+    fn update_idle(&self, event: IdleEvent) -> bool;
+    fn update_exit(&self, event: ExitEvent) -> bool;
+}
+
+mod Event {
+    pub struct Idle;
+    pub struct Exit;
+}
+
+use Event::*
+
+struct Object;
+impl Updateable<Idle, Exit> for Object {
+    fn update_idle(&self, event: Idle) -> bool {
+        true
+    }
+    fn update_exit(&self, event: Exit) -> bool {
+        false
+    }
+}
+
+fn run() {
+    let current_updatable = (&Object as &dyn Updateable<IdleEvent, ExitEvent>);
+    loop {
+        if recieve_exit_event() {
+            if !current_updatable.update_exit(Event::Exit) {
+                break;
+            }
+        }
+        if !current_updatable.update_idle(Event::Idle) {
+            break;
+        }
+    }
+}
+```
+
+This new approach will not create enum types just to match them directly afterwards, when it's still known, and removes a little unnecessary overhead.
+
+But this comes with problems. When adding a new event type, it has to be added in many places instead of just once in the enum declaration and once in the match. The match even will care if you forget the branch. So this small performance benefit is most likely not worth it.
```

Comment:
"The match even will care if you forget the branch" what does this mean?

Reply:
`match` in rust generates a warning, when the match is not exhausting, so when adding a new variant to the enum, you won't forget to add this branch in `match`. That's nothing new, but an advantage opposed to the more efficient approach I mentioned.
Maybe I should reformulate it this.